{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Environment Setup\n",
    "\n",
    "This notebook demonstrates how to genotype recurrent CNVs. The only required input files is:\n",
    "1. QCed CNV calls in .bed format: tab separated with header and first three columns: chrom, start, end.\n",
    "\n",
    "The required tools are:\n",
    "1. python (with pandas installed)\n",
    "2. bedtools (code to install static binary below)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-09-01 15:11:34 URL:https://release-assets.githubusercontent.com/github-production-release-asset/15059334/6cdff700-4d6a-11e9-9ddb-98618909f426?sp=r&sv=2018-11-09&sr=b&spr=https&se=2025-09-01T22%3A52%3A09Z&rscd=attachment%3B+filename%3Dbedtools&rsct=application%2Foctet-stream&skoid=96c2d410-5711-43a1-aedd-ab1947aa7ab0&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skt=2025-09-01T21%3A51%3A36Z&ske=2025-09-01T22%3A52%3A09Z&sks=b&skv=2018-11-09&sig=M%2FtuyOn2mzmhPGGjLEopcPlQlAkGqDwUSAJCI3009Cw%3D&jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmVsZWFzZS1hc3NldHMuZ2l0aHVidXNlcmNvbnRlbnQuY29tIiwia2V5Ijoia2V5MSIsImV4cCI6MTc1Njc2NDkzNCwibmJmIjoxNzU2NzY0NjM0LCJwYXRoIjoicmVsZWFzZWFzc2V0cHJvZHVjdGlvbi5ibG9iLmNvcmUud2luZG93cy5uZXQifQ.Qq-9p8Mtn7aN2TNC2Fxo_yyceR_tZbA8clmt_3e83pQ&response-content-disposition=attachment%3B%20filename%3Dbedtools&response-content-type=application%2Foctet-stream [38962368/38962368] -> \"bedtools\" [1]\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "wget -nv -O bedtools https://github.com/arq5x/bedtools2/releases/download/v2.28.0/bedtools\n",
    "chmod a+x bedtools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, set cohort name and date for saving output files. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "import os\n",
    "\n",
    "# Set variables\n",
    "COHORT = \"G2MH\"\n",
    "TODAY = datetime.date.today().strftime(\"%Y-%m-%d\")\n",
    "\n",
    "# Optionally, export to environment variables for shell cells\n",
    "os.environ[\"COHORT\"] = COHORT\n",
    "os.environ[\"TODAY\"] = TODAY\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read in and format locus definitions\n",
    "\n",
    "The locus definitions are defined in this public spreadsheet: https://docs.google.com/spreadsheets/d/1v6bzdeFYorIJsjuBDSTdMDHQVU47IrLI/edit?pli=1&gid=1816231862#gid=1816231862\n",
    "\n",
    "These blocks of code read in two sets of locus definitions:\n",
    "1. Loci that are genotyped with 50% reciprocal overlap\n",
    "2. Loci that are genotyped with 50% one-way overlap\n",
    "\n",
    "Regions that have multiple adjacent or nested breakpoints are genotyped with one-way overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hi\n",
      "2\t110105139\t110226371\t2q13_NPHP1\n",
      "2\t110636463\t111255072\t2q13\n",
      "2\t130723735\t131173104\t2q21.1\n",
      "7\t75508972\t76435095\t7q11.23_distal\n",
      "10\t48182156\t49850750\t10q11.21_11.23\n",
      "10\t80285716\t87171894\t10q22-23\n",
      "13\t22981219\t24310484\t13q12.12\n",
      "17\t14238070\t15523647\t17p12\n",
      "17\t16909457\t20307704\t17p11.2\n",
      "17\t36460090\t37857812\t17q12\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "sheet_id = \"1v6bzdeFYorIJsjuBDSTdMDHQVU47IrLI\"  # your sheet ID\n",
    "gid = \"1816231862\"  # sheet tab ID\n",
    "url_reciprocal = f'https://docs.google.com/spreadsheets/d/{sheet_id}/export?format=tsv&gid={gid}'\n",
    "print(\"hi\")\n",
    "df_reciprocal = pd.read_csv(url_reciprocal, sep=\"\\t\")\n",
    "\n",
    "bed_reciprocal = df_reciprocal.loc[:,['CHR', 'START', 'END', 'Locus name']]\n",
    "bed_reciprocal.to_csv('recurrentCNV_hg38.reciprocal.bed', index=False, sep='\\t', header=False)\n",
    "!tail recurrentCNV_hg38.reciprocal.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22\t20400000\t20665000\t22q11.21_B_C\n",
      "22\t20750000\t21112437\t22q11.21_C_D\n",
      "22\t21565838\t22610000\t22q11.21_D_E\n",
      "22\t22825000\t23311459\t22q11.21_E_F\n",
      "22\t23550000\t23950000\t22q11.21_F_G\n",
      "22\t24050000\t24200000\t22q11.21_G_H\n",
      "15\t30788442\t31724867\t15q13.3_BP4_BP5_exclCHRNA7\n",
      "16\t21570000\t21720000\t16p12.2_BP1_BP2\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\n",
      "16\t29035462\t29639519\t16p11.2_BP3_BP4\n"
     ]
    }
   ],
   "source": [
    "sheet_id = \"1v6bzdeFYorIJsjuBDSTdMDHQVU47IrLI\"  # your sheet ID\n",
    "gid = \"1406215423\"  # sheet tab ID\n",
    "url_oneway = f'https://docs.google.com/spreadsheets/d/{sheet_id}/export?format=tsv&gid={gid}'\n",
    "\n",
    "df_oneway = pd.read_csv(url_oneway, sep=\"\\t\")\n",
    "\n",
    "bed_oneway = df_oneway.loc[:,['CHR', 'START', 'END', 'Locus name']]\n",
    "bed_oneway.columns = ['locusCHR', 'locusSTART', 'locusEND', 'locusNAME']\n",
    "bed_oneway.to_csv('recurrentCNV_hg38.oneway.bed', index=False, sep='\\t', header=False)\n",
    "!tail recurrentCNV_hg38.oneway.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reformat CNV calls\n",
    "\n",
    "bedtools requires the first three columns of an input .bed file to be chromosome (column 1 in All Of Us \"chrom\"), start position (column 7, \"consStart\"), and end position (column 8, \"consEnd\"). We also need to include copy number (column 9, \"consCallType\") and sample ID (column 10, \"sampleId\").\n",
    "\n",
    "All Of Us CNV calls are split into batches in a GCS bucket. This next block of code extracts the relevant columns and concatenates each batch. It also removes the \"chr\" prefix from the chromosome column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\t48182475\t48674913\t.\t882-NDARAA241UP7\n",
      "10\t48674913\t48676662\t<DEL>\t882-NDARAA241UP7\n",
      "10\t48676663\t49795310\t.\t882-NDARAA241UP7\n",
      "10\t49795310\t49803465\t<DEL>\t882-NDARAA241UP7\n",
      "10\t49803466\t50064424\t.\t882-NDARAA241UP7\n",
      "10\t79880173\t87110741\t.\t882-NDARAA241UP7\n",
      "11\t41679753\t46139337\t.\t882-NDARAA241UP7\n",
      "13\t18786404\t22019381\t.\t882-NDARAA241UP7\n",
      "13\t22022972\t23058400\t.\t882-NDARAA241UP7\n",
      "13\t23058400\t23062975\t<DEL>\t882-NDARAA241UP7\n",
      "\n",
      "-rw-r--r-- 1 a1sriniv ddp195 4.2M Sep  1 15:11 /expanse/projects/sebat1/a1sriniv/g2mh/cnv_genotyping/wgs/other_recurrent_loci/G2MH.CNVcalls.consensus.2025-09-01.bed\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "output=\"/expanse/projects/sebat1/a1sriniv/g2mh/cnv_genotyping/wgs/other_recurrent_loci/${COHORT}.CNVcalls.consensus.${TODAY}.bed\"\n",
    ": > \"$output\"  # empty file\n",
    "\n",
    "for f in /expanse/projects/sebat1/a1sriniv/g2mh/cnv_genotyping/wgs/other_recurrent_loci/cnv_vcf_beds/*.bed; do\n",
    "    sample=\"${f##*/}\"                 # strip path → 882-NDARAA241UP7.filtered.bed\n",
    "    sample=\"${sample%.bed}\"           # strip .bed → 882-NDARAA241UP7.filtered\n",
    "    sample=\"${sample%.filtered}\" \n",
    "    awk -v sample=\"$sample\" -F \"\\t\" '{ sub(/^chr/, \"\", $1); print $1\"\\t\"$2\"\\t\"$3\"\\t\"$4\"\\t\"sample}' \"$f\"\n",
    "done >> $output\n",
    "\n",
    "head \"$output\"\n",
    "echo\n",
    "ls -lh \"$output\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run bedtools with 50%  reciprocal overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t147593490\t.\t883-NDARPK704YC5\n",
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t147725638\t.\t887-NDARBE180DFH\n",
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t148013556\t.\t882-NDARAA241UP7\n",
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t148013556\t.\t882-NDARAB886WJL\n",
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t148013556\t.\t882-NDARAC684ABP\n",
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t148013556\t.\t882-NDARAE674MBB\n",
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t148013556\t.\t882-NDARAF239LV0\n",
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t148013556\t.\t882-NDARAM990MEA\n",
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t148013556\t.\t882-NDARAN856ML7\n",
      "1\t147056425\t147922330\t1q21.1\t1\t147021747\t148013556\t.\t882-NDARAP035WH4\n",
      "-rw-r--r-- 1 a1sriniv ddp195 786K Sep  1 15:11 G2MH.reciprocal.2025-09-01.bed\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "./bedtools intersect \\\n",
    "      -a recurrentCNV_hg38.reciprocal.bed \\\n",
    "      -b ${COHORT}.CNVcalls.consensus.${TODAY}.bed\\\n",
    "      -header -wb -wa -f .5 -r > ${COHORT}.reciprocal.${TODAY}.bed\n",
    "      \n",
    "head ${COHORT}.reciprocal.${TODAY}.bed\n",
    "ls -lh ${COHORT}.reciprocal.${TODAY}.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read them into a pandas dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locusCHROM</th>\n",
       "      <th>locusSTART</th>\n",
       "      <th>locusEND</th>\n",
       "      <th>locusName</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>sampleID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>3</td>\n",
       "      <td>196011534</td>\n",
       "      <td>197617305</td>\n",
       "      <td>3q29</td>\n",
       "      <td>3</td>\n",
       "      <td>195595166</td>\n",
       "      <td>197237244</td>\n",
       "      <td>&lt;DUP&gt;</td>\n",
       "      <td>882-NDARVR804CER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1323</th>\n",
       "      <td>3</td>\n",
       "      <td>196011534</td>\n",
       "      <td>197617305</td>\n",
       "      <td>3q29</td>\n",
       "      <td>3</td>\n",
       "      <td>195750205</td>\n",
       "      <td>197157392</td>\n",
       "      <td>&lt;DUP&gt;</td>\n",
       "      <td>883-NDARVV291LUK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569</th>\n",
       "      <td>3</td>\n",
       "      <td>196011534</td>\n",
       "      <td>197617305</td>\n",
       "      <td>3q29</td>\n",
       "      <td>3</td>\n",
       "      <td>195261104</td>\n",
       "      <td>197235168</td>\n",
       "      <td>&lt;DUP&gt;</td>\n",
       "      <td>886-NDARRC407LCN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1606</th>\n",
       "      <td>3</td>\n",
       "      <td>196011534</td>\n",
       "      <td>197617305</td>\n",
       "      <td>3q29</td>\n",
       "      <td>3</td>\n",
       "      <td>195750205</td>\n",
       "      <td>197170422</td>\n",
       "      <td>&lt;DUP&gt;</td>\n",
       "      <td>886-NDARXN968MRG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1873</th>\n",
       "      <td>3</td>\n",
       "      <td>196011534</td>\n",
       "      <td>197617305</td>\n",
       "      <td>3q29</td>\n",
       "      <td>3</td>\n",
       "      <td>195276359</td>\n",
       "      <td>197235168</td>\n",
       "      <td>&lt;DUP&gt;</td>\n",
       "      <td>896-NDARDP871REA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      locusCHROM  locusSTART   locusEND locusName  CHROM      START  \\\n",
       "1197           3   196011534  197617305      3q29      3  195595166   \n",
       "1323           3   196011534  197617305      3q29      3  195750205   \n",
       "1569           3   196011534  197617305      3q29      3  195261104   \n",
       "1606           3   196011534  197617305      3q29      3  195750205   \n",
       "1873           3   196011534  197617305      3q29      3  195276359   \n",
       "\n",
       "            END   TYPE          sampleID  \n",
       "1197  197237244  <DUP>  882-NDARVR804CER  \n",
       "1323  197157392  <DUP>  883-NDARVV291LUK  \n",
       "1569  197235168  <DUP>  886-NDARRC407LCN  \n",
       "1606  197170422  <DUP>  886-NDARXN968MRG  \n",
       "1873  197235168  <DUP>  896-NDARDP871REA  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reciprocal_calls = pd.read_csv(f'{COHORT}.reciprocal.{TODAY}.bed', sep='\\t')\n",
    "header = ['locusCHROM', 'locusSTART', 'locusEND', 'locusName', 'CHROM', 'START', 'END', 'TYPE', 'sampleID']\n",
    "reciprocal_calls.columns = header\n",
    "reciprocal_calls = reciprocal_calls[reciprocal_calls['TYPE'] != '.']\n",
    "reciprocal_calls.head()\n",
    "#reciprocal_calls['TYPE'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run bedtools with 50% one-way overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t24481230\t28318191\t.\t896-NDARYX075FCC\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t23305673\t28318191\t.\t896-NDARZG817TV3\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t23815959\t28318191\t.\t896-NDARZH271GK6\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t24874491\t28293502\t.\t896-NDARZH361EM0\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t24874491\t28318191\t.\t896-NDARZH390XM2\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t23305673\t28318191\t.\t896-NDARZJ463MTD\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t24874491\t28318191\t.\t896-NDARZV016GBP\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t23750645\t28318191\t.\t896-NDARZW083WJ8\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t23305673\t28318191\t.\t896-NDARZW810BU0\n",
      "15\t23067754\t28145193\t15q11.2-13.1_BP2_BP3\t15\t23305673\t25972093\t.\t896-NDARZX589WUG\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "./bedtools intersect \\\n",
    "      -a recurrentCNV_hg38.oneway.bed \\\n",
    "      -b ${COHORT}.CNVcalls.consensus.${TODAY}.bed\\\n",
    "      -header -wb -wa -f .5 > ${COHORT}.oneway.${TODAY}.bed\n",
    "      \n",
    "tail ${COHORT}.oneway.${TODAY}.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read them into a pandas dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "locusName\n",
       "22q11.21_A_B                  705\n",
       "22q11.21_C_D                  684\n",
       "22q11.21_B_C                  680\n",
       "16p11.2_BP4_BP5               226\n",
       "16p11.2_BP2_BP3                55\n",
       "22q11.21_D_E                   34\n",
       "22q11.21_E_F                   29\n",
       "22q11.21_F_G                   26\n",
       "22q11.21_G_H                   25\n",
       "16p11.2_BP1_BP2                16\n",
       "15q11.2_BP1_BP2                14\n",
       "15q13.3_CHRNA7                 11\n",
       "1q21.1_TAR_BP1_BP2              8\n",
       "1q21.1_TAR_BP2_BP3              6\n",
       "16p13.11_BP1_BP2                6\n",
       "15q13.3_BP4_BP5_exclCHRNA7      6\n",
       "16p12.2_BP2_BP3                 4\n",
       "15q13.1-13.2_BP3_BP4            2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oneway_calls = pd.read_csv(f'{COHORT}.oneway.{TODAY}.bed', sep='\\t')\n",
    "header = ['locusCHROM', 'locusSTART', 'locusEND', 'locusName', 'CHROM', 'START', 'END', 'TYPE', 'sampleID']\n",
    "oneway_calls.columns = header\n",
    "oneway_calls = oneway_calls[oneway_calls['TYPE'] != '.']\n",
    "oneway_calls['locusName'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Format calls into genotype matrix\n",
    "\n",
    "For All Of Us, I initialized this matrix by reading in the manifest that includes sample QC ('Freq_Calc' column, formatted 1 for pass, 0 for fail). For other datasets, any matrix that contains one row for each sample ID (i.e. a sample manifest) works. Set the 'identifier' column to be the sampleIDs that match up with the CNV calls\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GUID</th>\n",
       "      <th>Genomic_ID</th>\n",
       "      <th>nda_dataset</th>\n",
       "      <th>nda_id</th>\n",
       "      <th>comment</th>\n",
       "      <th>WGS.GSA</th>\n",
       "      <th>guid</th>\n",
       "      <th>redcap_data_access_group</th>\n",
       "      <th>age_final</th>\n",
       "      <th>sex_final</th>\n",
       "      <th>...</th>\n",
       "      <th>genomics_sample_type</th>\n",
       "      <th>rutgers_status</th>\n",
       "      <th>tcag_status</th>\n",
       "      <th>gen_status_nda</th>\n",
       "      <th>final_group_class</th>\n",
       "      <th>final_group_22q_type</th>\n",
       "      <th>final_group_16p_type</th>\n",
       "      <th>group_class_breakpoints</th>\n",
       "      <th>enroll_year</th>\n",
       "      <th>identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NDARAA121WTA</td>\n",
       "      <td>887-NDARAA121WTA</td>\n",
       "      <td>Jul_2025_WGS</td>\n",
       "      <td>75704.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAA121WTA</td>\n",
       "      <td>toronto</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Duplication</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Duplication A-D</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>887-NDARAA121WTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NDARAA241UP7</td>\n",
       "      <td>882-NDARAA241UP7</td>\n",
       "      <td>Jul_2025_WGS</td>\n",
       "      <td>75704.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAA241UP7</td>\n",
       "      <td>maastricht</td>\n",
       "      <td>31.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Deletion</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Deletion A-D</td>\n",
       "      <td>2024.0</td>\n",
       "      <td>882-NDARAA241UP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NDARAA984CN8</td>\n",
       "      <td>896-NDARAA984CN8</td>\n",
       "      <td>Jul_2023_WGS</td>\n",
       "      <td>59392.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAA984CN8</td>\n",
       "      <td>pennchop</td>\n",
       "      <td>44.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22q11.2_Duplication</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Duplication A-D</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>896-NDARAA984CN8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NDARAB126ULP</td>\n",
       "      <td>884-NDARAB126ULP</td>\n",
       "      <td>Jan_2024_WGS</td>\n",
       "      <td>64244.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAB126ULP</td>\n",
       "      <td>ucla</td>\n",
       "      <td>9.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22q11.2_Deletion</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Deletion A-D</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>884-NDARAB126ULP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NDARAB421WYL</td>\n",
       "      <td>883-NDARAB421WYL</td>\n",
       "      <td>Jul_2023_WGS</td>\n",
       "      <td>59392.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAB421WYL</td>\n",
       "      <td>leuven</td>\n",
       "      <td>33.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22q11.2_Deletion</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Deletion A-D</td>\n",
       "      <td>2022.0</td>\n",
       "      <td>883-NDARAB421WYL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2417</th>\n",
       "      <td>NDARZX589WUG</td>\n",
       "      <td>896-NDARZX589WUG</td>\n",
       "      <td>Jul_2023_WGS</td>\n",
       "      <td>59392.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARZX589WUG</td>\n",
       "      <td>pennchop</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22q11.2_Deletion</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Deletion A-D</td>\n",
       "      <td>2021.0</td>\n",
       "      <td>896-NDARZX589WUG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2419</th>\n",
       "      <td>NDARZX886ZDC</td>\n",
       "      <td>896-NDARZX886ZDC</td>\n",
       "      <td>Jan_2023</td>\n",
       "      <td>53739.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARZX886ZDC</td>\n",
       "      <td>pennchop</td>\n",
       "      <td>15.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22q11.2_Deletion</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Deletion A-D</td>\n",
       "      <td>2020.0</td>\n",
       "      <td>896-NDARZX886ZDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2420</th>\n",
       "      <td>NDARZX911VCJ</td>\n",
       "      <td>884-NDARZX911VCJ</td>\n",
       "      <td>Jan_2023</td>\n",
       "      <td>53739.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARZX911VCJ</td>\n",
       "      <td>ucla</td>\n",
       "      <td>10.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22q11.2_Duplication</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Duplication A-D</td>\n",
       "      <td>2021.0</td>\n",
       "      <td>884-NDARZX911VCJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2428</th>\n",
       "      <td>NDARZZ354KTJ</td>\n",
       "      <td>888-NDARZZ354KTJ</td>\n",
       "      <td>Jan_2024_WGS</td>\n",
       "      <td>64244.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARZZ354KTJ</td>\n",
       "      <td>montreal</td>\n",
       "      <td>20.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>22q11.2_Duplication</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Duplication A-D</td>\n",
       "      <td>2021.0</td>\n",
       "      <td>888-NDARZZ354KTJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2429</th>\n",
       "      <td>NDARZZ419FG3</td>\n",
       "      <td>896-NDARZZ419FG3</td>\n",
       "      <td>Jul_2025_WGS</td>\n",
       "      <td>75704.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARZZ419FG3</td>\n",
       "      <td>pennchop</td>\n",
       "      <td>13.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Deletion</td>\n",
       "      <td>A-D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22q11.2_Deletion A-D</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>896-NDARZZ419FG3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1085 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              GUID        Genomic_ID   nda_dataset   nda_id comment WGS.GSA  \\\n",
       "0     NDARAA121WTA  887-NDARAA121WTA  Jul_2025_WGS  75704.0     NaN     WGS   \n",
       "1     NDARAA241UP7  882-NDARAA241UP7  Jul_2025_WGS  75704.0     NaN     WGS   \n",
       "5     NDARAA984CN8  896-NDARAA984CN8  Jul_2023_WGS  59392.0     NaN     WGS   \n",
       "8     NDARAB126ULP  884-NDARAB126ULP  Jan_2024_WGS  64244.0     NaN     WGS   \n",
       "9     NDARAB421WYL  883-NDARAB421WYL  Jul_2023_WGS  59392.0     NaN     WGS   \n",
       "...            ...               ...           ...      ...     ...     ...   \n",
       "2417  NDARZX589WUG  896-NDARZX589WUG  Jul_2023_WGS  59392.0     NaN     WGS   \n",
       "2419  NDARZX886ZDC  896-NDARZX886ZDC      Jan_2023  53739.0     NaN     WGS   \n",
       "2420  NDARZX911VCJ  884-NDARZX911VCJ      Jan_2023  53739.0     NaN     WGS   \n",
       "2428  NDARZZ354KTJ  888-NDARZZ354KTJ  Jan_2024_WGS  64244.0     NaN     WGS   \n",
       "2429  NDARZZ419FG3  896-NDARZZ419FG3  Jul_2025_WGS  75704.0     NaN     WGS   \n",
       "\n",
       "              guid redcap_data_access_group  age_final sex_final  ...  \\\n",
       "0     NDARAA121WTA                  toronto        7.0    Female  ...   \n",
       "1     NDARAA241UP7               maastricht       31.0      Male  ...   \n",
       "5     NDARAA984CN8                 pennchop       44.0      Male  ...   \n",
       "8     NDARAB126ULP                     ucla        9.0    Female  ...   \n",
       "9     NDARAB421WYL                   leuven       33.0    Female  ...   \n",
       "...            ...                      ...        ...       ...  ...   \n",
       "2417  NDARZX589WUG                 pennchop       24.0    Female  ...   \n",
       "2419  NDARZX886ZDC                 pennchop       15.0    Female  ...   \n",
       "2420  NDARZX911VCJ                     ucla       10.0      Male  ...   \n",
       "2428  NDARZZ354KTJ                 montreal       20.0      Male  ...   \n",
       "2429  NDARZZ419FG3                 pennchop       13.0      Male  ...   \n",
       "\n",
       "     genomics_sample_type rutgers_status tcag_status  gen_status_nda  \\\n",
       "0                     1.0            2.0         NaN             NaN   \n",
       "1                     2.0            NaN         1.0             NaN   \n",
       "5                     2.0            NaN         1.0             1.0   \n",
       "8                     1.0            2.0         NaN             1.0   \n",
       "9                     1.0            NaN         1.0             1.0   \n",
       "...                   ...            ...         ...             ...   \n",
       "2417                  1.0            NaN         1.0             1.0   \n",
       "2419                  1.0            NaN         1.0             1.0   \n",
       "2420                  2.0            2.0         NaN             1.0   \n",
       "2428                  1.0            2.0         NaN             1.0   \n",
       "2429                  NaN            NaN         NaN             NaN   \n",
       "\n",
       "        final_group_class final_group_22q_type final_group_16p_type  \\\n",
       "0     22q11.2_Duplication                  A-D                  NaN   \n",
       "1        22q11.2_Deletion                  A-D                  NaN   \n",
       "5     22q11.2_Duplication                  A-D                  NaN   \n",
       "8        22q11.2_Deletion                  A-D                  NaN   \n",
       "9        22q11.2_Deletion                  A-D                  NaN   \n",
       "...                   ...                  ...                  ...   \n",
       "2417     22q11.2_Deletion                  A-D                  NaN   \n",
       "2419     22q11.2_Deletion                  A-D                  NaN   \n",
       "2420  22q11.2_Duplication                  A-D                  NaN   \n",
       "2428  22q11.2_Duplication                  A-D                  NaN   \n",
       "2429     22q11.2_Deletion                  A-D                  NaN   \n",
       "\n",
       "      group_class_breakpoints enroll_year        identifier  \n",
       "0     22q11.2_Duplication A-D      2023.0  887-NDARAA121WTA  \n",
       "1        22q11.2_Deletion A-D      2024.0  882-NDARAA241UP7  \n",
       "5     22q11.2_Duplication A-D      2022.0  896-NDARAA984CN8  \n",
       "8        22q11.2_Deletion A-D      2022.0  884-NDARAB126ULP  \n",
       "9        22q11.2_Deletion A-D      2022.0  883-NDARAB421WYL  \n",
       "...                       ...         ...               ...  \n",
       "2417     22q11.2_Deletion A-D      2021.0  896-NDARZX589WUG  \n",
       "2419     22q11.2_Deletion A-D      2020.0  896-NDARZX886ZDC  \n",
       "2420  22q11.2_Duplication A-D      2021.0  884-NDARZX911VCJ  \n",
       "2428  22q11.2_Duplication A-D      2021.0  888-NDARZZ354KTJ  \n",
       "2429     22q11.2_Deletion A-D      2023.0  896-NDARZZ419FG3  \n",
       "\n",
       "[1085 rows x 29 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampleQC = 'full_sample_manifest.csv'\n",
    "genomatrix = pd.read_csv(sampleQC, sep=\",\")\n",
    "genomatrix = genomatrix[genomatrix['WGS.GSA'] == 'WGS']\n",
    "genomatrix['identifier'] = genomatrix['Genomic_ID']\n",
    "genomatrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, add reciprocal calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map CNV types to numeric codes\n",
    "cn_mapping = {'<DEL>': 1, '<DUP>': 3}\n",
    "calls_df = reciprocal_calls.copy()\n",
    "calls_df['CN'] = calls_df['TYPE'].map(cn_mapping)\n",
    "\n",
    "# Get all unique loci\n",
    "loci = calls_df['locusName'].unique()\n",
    "loci_reciprocal = loci\n",
    "\n",
    "# Ensure 'identifier' and 'sampleID' are string type for consistent merging\n",
    "genomatrix['identifier'] = genomatrix['identifier'].astype(str)\n",
    "calls_df['sampleID'] = calls_df['sampleID'].astype(str)\n",
    "\n",
    "# Initialize columns in genomatrix_CNV with default value 2\n",
    "for l in loci:\n",
    "    genomatrix[l] = 2\n",
    "\n",
    "# Pivot CNV calls to wide format (sample x locus)\n",
    "pivoted = calls_df.pivot_table(index='sampleID', columns='locusName', values='CN', aggfunc='first')\n",
    "\n",
    "# Merge CNV calls into genotype matrix using update\n",
    "pivoted.index.name = 'identifier'  # match index name to merge key\n",
    "genomatrix.set_index('identifier', inplace=True)\n",
    "genomatrix.update(pivoted)\n",
    "genomatrix.reset_index(inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, add one-way calls. We add the suffix \"oneway\" to the locus names to indicate that these are not yet \"allele specific\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>GUID</th>\n",
       "      <th>Genomic_ID</th>\n",
       "      <th>nda_dataset</th>\n",
       "      <th>nda_id</th>\n",
       "      <th>comment</th>\n",
       "      <th>WGS.GSA</th>\n",
       "      <th>guid</th>\n",
       "      <th>redcap_data_access_group</th>\n",
       "      <th>age_final</th>\n",
       "      <th>...</th>\n",
       "      <th>16p11.2_BP2_BP3_oneway</th>\n",
       "      <th>16p11.2_BP4_BP5_oneway</th>\n",
       "      <th>22q11.21_A_B_oneway</th>\n",
       "      <th>22q11.21_B_C_oneway</th>\n",
       "      <th>22q11.21_C_D_oneway</th>\n",
       "      <th>22q11.21_D_E_oneway</th>\n",
       "      <th>22q11.21_E_F_oneway</th>\n",
       "      <th>22q11.21_F_G_oneway</th>\n",
       "      <th>22q11.21_G_H_oneway</th>\n",
       "      <th>15q13.3_BP4_BP5_exclCHRNA7_oneway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>887-NDARAA121WTA</td>\n",
       "      <td>NDARAA121WTA</td>\n",
       "      <td>887-NDARAA121WTA</td>\n",
       "      <td>Jul_2025_WGS</td>\n",
       "      <td>75704.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAA121WTA</td>\n",
       "      <td>toronto</td>\n",
       "      <td>7.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>882-NDARAA241UP7</td>\n",
       "      <td>NDARAA241UP7</td>\n",
       "      <td>882-NDARAA241UP7</td>\n",
       "      <td>Jul_2025_WGS</td>\n",
       "      <td>75704.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAA241UP7</td>\n",
       "      <td>maastricht</td>\n",
       "      <td>31.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>896-NDARAA984CN8</td>\n",
       "      <td>NDARAA984CN8</td>\n",
       "      <td>896-NDARAA984CN8</td>\n",
       "      <td>Jul_2023_WGS</td>\n",
       "      <td>59392.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAA984CN8</td>\n",
       "      <td>pennchop</td>\n",
       "      <td>44.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>884-NDARAB126ULP</td>\n",
       "      <td>NDARAB126ULP</td>\n",
       "      <td>884-NDARAB126ULP</td>\n",
       "      <td>Jan_2024_WGS</td>\n",
       "      <td>64244.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAB126ULP</td>\n",
       "      <td>ucla</td>\n",
       "      <td>9.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>883-NDARAB421WYL</td>\n",
       "      <td>NDARAB421WYL</td>\n",
       "      <td>883-NDARAB421WYL</td>\n",
       "      <td>Jul_2023_WGS</td>\n",
       "      <td>59392.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WGS</td>\n",
       "      <td>NDARAB421WYL</td>\n",
       "      <td>leuven</td>\n",
       "      <td>33.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         identifier          GUID        Genomic_ID   nda_dataset   nda_id  \\\n",
       "0  887-NDARAA121WTA  NDARAA121WTA  887-NDARAA121WTA  Jul_2025_WGS  75704.0   \n",
       "1  882-NDARAA241UP7  NDARAA241UP7  882-NDARAA241UP7  Jul_2025_WGS  75704.0   \n",
       "2  896-NDARAA984CN8  NDARAA984CN8  896-NDARAA984CN8  Jul_2023_WGS  59392.0   \n",
       "3  884-NDARAB126ULP  NDARAB126ULP  884-NDARAB126ULP  Jan_2024_WGS  64244.0   \n",
       "4  883-NDARAB421WYL  NDARAB421WYL  883-NDARAB421WYL  Jul_2023_WGS  59392.0   \n",
       "\n",
       "  comment WGS.GSA          guid redcap_data_access_group  age_final  ...  \\\n",
       "0     NaN     WGS  NDARAA121WTA                  toronto        7.0  ...   \n",
       "1     NaN     WGS  NDARAA241UP7               maastricht       31.0  ...   \n",
       "2     NaN     WGS  NDARAA984CN8                 pennchop       44.0  ...   \n",
       "3     NaN     WGS  NDARAB126ULP                     ucla        9.0  ...   \n",
       "4     NaN     WGS  NDARAB421WYL                   leuven       33.0  ...   \n",
       "\n",
       "  16p11.2_BP2_BP3_oneway 16p11.2_BP4_BP5_oneway 22q11.21_A_B_oneway  \\\n",
       "0                      2                      2                   3   \n",
       "1                      2                      2                   1   \n",
       "2                      2                      2                   3   \n",
       "3                      2                      2                   1   \n",
       "4                      2                      2                   1   \n",
       "\n",
       "  22q11.21_B_C_oneway  22q11.21_C_D_oneway 22q11.21_D_E_oneway  \\\n",
       "0                   3                    3                   2   \n",
       "1                   1                    1                   2   \n",
       "2                   3                    3                   2   \n",
       "3                   1                    1                   2   \n",
       "4                   1                    1                   2   \n",
       "\n",
       "  22q11.21_E_F_oneway 22q11.21_F_G_oneway 22q11.21_G_H_oneway  \\\n",
       "0                   2                   2                   2   \n",
       "1                   2                   2                   2   \n",
       "2                   2                   2                   2   \n",
       "3                   2                   2                   2   \n",
       "4                   2                   2                   2   \n",
       "\n",
       "  15q13.3_BP4_BP5_exclCHRNA7_oneway  \n",
       "0                                 2  \n",
       "1                                 2  \n",
       "2                                 2  \n",
       "3                                 2  \n",
       "4                                 2  \n",
       "\n",
       "[5 rows x 57 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map CNV types to numeric codes\n",
    "cn_mapping = {'<DEL>': 1, '<DUP>': 3}\n",
    "calls_df = oneway_calls.copy()\n",
    "calls_df['CN'] = calls_df['TYPE'].map(cn_mapping)\n",
    "\n",
    "# Get all unique loci\n",
    "loci = calls_df['locusName'].unique()\n",
    "loci_oneway = [f'{l}_oneway' for l in loci]\n",
    "\n",
    "# Ensure 'identifier' and 'sampleID' are string type for consistent merging\n",
    "genomatrix['identifier'] = genomatrix['identifier'].astype(str)\n",
    "calls_df['sampleID'] = calls_df['sampleID'].astype(str)\n",
    "calls_df['locusName'] = calls_df['locusName'] + '_oneway'\n",
    "\n",
    "# Initialize columns in genomatrix_CNV with default value 2\n",
    "for l in loci:\n",
    "    genomatrix[f'{l}_oneway'] = 2\n",
    "\n",
    "# Pivot CNV calls to wide format (sample x locus)\n",
    "pivoted = calls_df.pivot_table(index='sampleID', columns='locusName', values='CN', aggfunc='first')\n",
    "\n",
    "# Merge CNV calls into genotype matrix using update\n",
    "pivoted.index.name = 'identifier'  # match index name to merge key\n",
    "genomatrix.set_index('identifier', inplace=True)\n",
    "genomatrix.update(pivoted)\n",
    "genomatrix.reset_index(inplace=True)\n",
    "genomatrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15q13.3_BP4_BP5_exclCHRNA7_oneway\n",
       "2    1079\n",
       "3       3\n",
       "1       3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomatrix['15q13.3_BP4_BP5_exclCHRNA7_oneway'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['identifier', 'GUID', 'Genomic_ID', 'nda_dataset', 'nda_id', 'comment',\n",
       "       'WGS.GSA', 'guid', 'redcap_data_access_group', 'age_final', 'sex_final',\n",
       "       'gen_group_class', 'group_class', 'gen_group_16p_type',\n",
       "       'group_16p_type', 'gen_group_22q_type', 'group_22q_type',\n",
       "       'gen_group_nonstandard_type', 'gen_group_nonstandard_type_breakpoints',\n",
       "       'group_22q_type_other', 'genomics_sample_type', 'rutgers_status',\n",
       "       'tcag_status', 'gen_status_nda', 'final_group_class',\n",
       "       'final_group_22q_type', 'final_group_16p_type',\n",
       "       'group_class_breakpoints', 'enroll_year', '3q29', '7q11.23_WBS',\n",
       "       '2q13_NPHP1', '2q13', '2q21.1', '7q11.23_distal', '10q11.21_11.23',\n",
       "       '13q12.12', '17p12', '17q12', '1q21.1_TAR_BP1_BP2_oneway',\n",
       "       '1q21.1_TAR_BP2_BP3_oneway', '15q11.2_BP1_BP2_oneway',\n",
       "       '15q13.1-13.2_BP3_BP4_oneway', '15q13.3_CHRNA7_oneway',\n",
       "       '16p13.11_BP1_BP2_oneway', '16p12.2_BP2_BP3_oneway',\n",
       "       '16p11.2_BP1_BP2_oneway', '16p11.2_BP2_BP3_oneway',\n",
       "       '16p11.2_BP4_BP5_oneway', '22q11.21_A_B_oneway', '22q11.21_B_C_oneway',\n",
       "       '22q11.21_C_D_oneway', '22q11.21_D_E_oneway', '22q11.21_E_F_oneway',\n",
       "       '22q11.21_F_G_oneway', '22q11.21_G_H_oneway',\n",
       "       '15q13.3_BP4_BP5_exclCHRNA7_oneway'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomatrix.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# (Optional) set calls for samples that failed QC to NA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "all_loci = list(loci_oneway) + list(loci_reciprocal)\n",
    "#genomatrix.loc[genomatrix[\"NOTES..additional.filter.reason.\"] != 'pass', all_loci] = np.nan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Combine one-way calls into \"allele specific\" genotypes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1q21.1 TAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>1q21.1_TAR_BP1_BP2</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1q21.1_TAR_BP2_BP3</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1083</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "1q21.1_TAR_BP1_BP2     2  3\n",
       "1q21.1_TAR_BP2_BP3         \n",
       "2                   1083  2"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomatrix['1q21.1_TAR_BP1_BP2'] = np.where(\n",
    "    genomatrix['1q21.1_TAR_BP2_BP3_oneway'] == 2,\n",
    "    genomatrix['1q21.1_TAR_BP1_BP2_oneway'],\n",
    "    2\n",
    ")\n",
    "\n",
    "genomatrix['1q21.1_TAR_BP2_BP3'] = np.where(\n",
    "    genomatrix['1q21.1_TAR_BP1_BP2_oneway'] == 2,\n",
    "    genomatrix['1q21.1_TAR_BP2_BP3_oneway'],\n",
    "    2\n",
    ")\n",
    "\n",
    "genomatrix['1q21.1_TAR_BP1_BP3'] = np.where(\n",
    "    genomatrix['1q21.1_TAR_BP1_BP2_oneway'] == genomatrix['1q21.1_TAR_BP2_BP3_oneway'],\n",
    "    genomatrix['1q21.1_TAR_BP2_BP3_oneway'],\n",
    "    2\n",
    ")\n",
    "\n",
    "pd.crosstab(genomatrix['1q21.1_TAR_BP2_BP3'], genomatrix['1q21.1_TAR_BP1_BP2'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 15q11-13\n",
    "\n",
    "This region has two separate CNV regions: BP1-3 and BP3-5. CNVs at CYFIP and CHRNA7 are common ( ~ 0.5% AF), so it is not unlikely that there are subjects who have one of these CNVs AND an independent CNV at the other end of the region. So, we genotype BP1-BP3 and BP3-BP5 independently"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>15q11.2-13.1_BP2_BP3</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15q11.2_BP1_BP2</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "15q11.2-13.1_BP2_BP3     2\n",
       "15q11.2_BP1_BP2           \n",
       "1                        4\n",
       "2                     1071\n",
       "3                       10"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adding because g2mh does not have:\n",
    "genomatrix['15q11.2-13.1_BP2_BP3_oneway'] = 2\n",
    "\n",
    "regions_15q13 = ['15q11.2_BP1_BP2_oneway', \n",
    "               '15q11.2-13.1_BP2_BP3_oneway']\n",
    "genomatrix['15q13'] = (\n",
    "    genomatrix[regions_15q13]\n",
    "    .fillna(0)\n",
    "    .astype(int)\n",
    "    .astype(str)\n",
    "    .agg(''.join, axis=1)\n",
    ")\n",
    "\n",
    "## 1-2 CYFIP\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['15q13'] == '12',\n",
    "    genomatrix['15q13'] == '32',\n",
    "    genomatrix['15q13'] == '00'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['15q11.2_BP1_BP2'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## 2-3 \n",
    "\n",
    "conditions = [\n",
    "    genomatrix['15q13'] == '21',\n",
    "    genomatrix['15q13'] == '23',\n",
    "    genomatrix['15q13'] == '00'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['15q11.2-13.1_BP2_BP3'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## 1-3 \n",
    "\n",
    "conditions = [\n",
    "    genomatrix['15q13'] == '11',\n",
    "    genomatrix['15q13'] == '33',\n",
    "    genomatrix['15q13'] == '00'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['15q11.2-13.1_BP1_BP3'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "pd.crosstab(genomatrix['15q11.2_BP1_BP2'], genomatrix['15q11.2-13.1_BP2_BP3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>15q13.3_BP4_BP5</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15q13.3_CHRNA7</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1075</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "15q13.3_BP4_BP5  1     2  3\n",
       "15q13.3_CHRNA7             \n",
       "2                3  1075  2\n",
       "3                0     5  0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions_15q35 = ['15q13.1-13.2_BP3_BP4_oneway',\n",
    "               '15q13.3_BP4_BP5_exclCHRNA7_oneway',\n",
    "               '15q13.3_CHRNA7_oneway']\n",
    "genomatrix['15q35'] = (\n",
    "    genomatrix[regions_15q35]\n",
    "    .fillna(0)\n",
    "    .astype(int)\n",
    "    .astype(str)\n",
    "    .agg(''.join, axis=1)\n",
    ")\n",
    "\n",
    "## 3-4 \n",
    "\n",
    "conditions = [\n",
    "    genomatrix['15q35'] == '122',\n",
    "    genomatrix['15q35'] == '322',\n",
    "    genomatrix['15q35'] == '000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['15q13.1-13.2_BP3_BP4'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## 3-5 \n",
    "\n",
    "conditions = [\n",
    "    genomatrix['15q35'] == '111',\n",
    "    genomatrix['15q35'] == '333',\n",
    "    genomatrix['15q35'] == '000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['15q13.1-13.3_BP3_BP5'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## 4-5 \n",
    "\n",
    "conditions = [\n",
    "    genomatrix['15q35'] == '211',\n",
    "    genomatrix['15q35'] == '233',\n",
    "    genomatrix['15q35'] == '000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['15q13.3_BP4_BP5'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## 4-5 excl CHRNA7\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['15q35'] == '212',\n",
    "    genomatrix['15q35'] == '232',\n",
    "    genomatrix['15q35'] == '000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['15q13.3_BP4_BP5_exclCHRNA7'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "## CHRNA7 \n",
    "\n",
    "conditions = [\n",
    "    genomatrix['15q35'] == '221',\n",
    "    genomatrix['15q35'] == '223',\n",
    "    genomatrix['15q35'] == '000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['15q13.3_CHRNA7'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "pd.crosstab(genomatrix['15q13.3_CHRNA7'], genomatrix['15q13.3_BP4_BP5'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 16p11.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>16p11.2_BP2_BP3</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16p11.2_BP4_BP5</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>145</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>28</td>\n",
       "      <td>828</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>73</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "16p11.2_BP2_BP3   1    2   3\n",
       "16p11.2_BP4_BP5             \n",
       "1                 0  145   0\n",
       "2                28  828  11\n",
       "3                 0   73   0"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomatrix['16p11.2_BP3_BP4_oneway'] = 2\n",
    "regions_16p11 = ['16p11.2_BP1_BP2_oneway', \n",
    "                 '16p11.2_BP2_BP3_oneway', \n",
    "                 '16p11.2_BP3_BP4_oneway',\n",
    "                 '16p11.2_BP4_BP5_oneway']\n",
    "genomatrix['16p11.2'] = (\n",
    "    genomatrix[regions_16p11]\n",
    "    .fillna(0)\n",
    "    .astype('int8')   # much smaller memory footprint\n",
    "    .astype(str)\n",
    "    .agg(''.join, axis=1)\n",
    ")\n",
    "\n",
    "## 1-2\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['16p11.2'] == '1222',\n",
    "    genomatrix['16p11.2'] == '3222',\n",
    "    genomatrix['16p11.2'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['16p11.2_BP1_BP2'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## 1-3\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['16p11.2'] == '1122',\n",
    "    genomatrix['16p11.2'] == '3322',\n",
    "    genomatrix['16p11.2'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['16p11.2_BP1_BP3'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## 2-3\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['16p11.2'] == '2122',\n",
    "    genomatrix['16p11.2'] == '2322',\n",
    "    genomatrix['16p11.2'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['16p11.2_BP2_BP3'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## 3-4\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['16p11.2'] == '2212',\n",
    "    genomatrix['16p11.2'] == '2232',\n",
    "    genomatrix['16p11.2'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['16p11.2_BP3_BP4'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "## 4-5\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['16p11.2'] == '2221',\n",
    "    genomatrix['16p11.2'] == '2223',\n",
    "    genomatrix['16p11.2'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['16p11.2_BP4_BP5'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "pd.crosstab(genomatrix['16p11.2_BP4_BP5'], genomatrix['16p11.2_BP2_BP3'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 16p12.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>16p12.2_BP2_BP3</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16p12.2_BP1_BP2</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1081</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "16p12.2_BP2_BP3  1     2  3\n",
       "16p12.2_BP1_BP2            \n",
       "2                1  1081  3"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print(genomatrix.columns)\n",
    "genomatrix['16p12.2_BP1_BP2_oneway'] = 2\n",
    "\n",
    "genomatrix['16p12.2_BP1_BP2'] = np.where(\n",
    "    genomatrix['16p12.2_BP2_BP3_oneway'] == 2,\n",
    "    genomatrix['16p12.2_BP1_BP2_oneway'],\n",
    "    2\n",
    ")\n",
    "\n",
    "genomatrix['16p12.2_BP2_BP3'] = np.where(\n",
    "    genomatrix['16p12.2_BP1_BP2_oneway'] == 2,\n",
    "    genomatrix['16p12.2_BP2_BP3_oneway'],\n",
    "    2\n",
    ")\n",
    "\n",
    "genomatrix['16p12.2_BP1_BP3'] = np.where(\n",
    "    genomatrix['16p12.2_BP1_BP2_oneway'] == genomatrix['16p12.2_BP2_BP3_oneway'],\n",
    "    genomatrix['16p12.2_BP1_BP2_oneway'],\n",
    "    2\n",
    ")\n",
    "\n",
    "pd.crosstab(genomatrix['16p12.2_BP1_BP2'], genomatrix['16p12.2_BP2_BP3'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 16p13.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>16p13.11_BP2_BP3</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16p13.11_BP1_BP2</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "16p13.11_BP2_BP3     2\n",
       "16p13.11_BP1_BP2      \n",
       "1                    4\n",
       "2                 1079\n",
       "3                    2"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomatrix['16p13.11_BP2_BP3_oneway'] = 2\n",
    "\n",
    "genomatrix['16p13.11_BP1_BP2'] = np.where(\n",
    "    genomatrix['16p13.11_BP2_BP3_oneway'] == 2,\n",
    "    genomatrix['16p13.11_BP1_BP2_oneway'],\n",
    "    2\n",
    ")\n",
    "\n",
    "genomatrix['16p13.11_BP2_BP3'] = np.where(\n",
    "    genomatrix['16p13.11_BP1_BP2_oneway'] == 2,\n",
    "    genomatrix['16p13.11_BP2_BP3_oneway'],\n",
    "    2\n",
    ")\n",
    "\n",
    "genomatrix['16p13.11_BP1_BP3'] = np.where(\n",
    "    genomatrix['16p13.11_BP1_BP2_oneway'] == genomatrix['16p13.11_BP2_BP3_oneway'],\n",
    "    genomatrix['16p13.11_BP1_BP2_oneway'],\n",
    "    2\n",
    ")\n",
    "\n",
    "pd.crosstab(genomatrix['16p13.11_BP1_BP2'], genomatrix['16p13.11_BP2_BP3'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 22q11.2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, genotype the A-D region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>22q11.21_C_D</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22q11.21_A_D</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>501</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>423</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>141</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "22q11.21_C_D   1    2   3\n",
       "22q11.21_A_D             \n",
       "1              0  501   0\n",
       "2             10  423  10\n",
       "3              0  141   0"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions_22qAD = ['22q11.21_A_B_oneway', '22q11.21_B_C_oneway', '22q11.21_C_D_oneway']\n",
    "genomatrix['22q11.21_AD'] = (\n",
    "    genomatrix[regions_22qAD]\n",
    "    .fillna(0)\n",
    "    .astype(int)\n",
    "    .astype(str)\n",
    "    .agg(''.join, axis=1)\n",
    ")\n",
    "\n",
    "## AB\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_AD'] == '122',\n",
    "    genomatrix['22q11.21_AD'] == '322',\n",
    "    genomatrix['22q11.21_AD'] == '000',\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['22q11.21_A_B'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## BC\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_AD'] == '212',\n",
    "    genomatrix['22q11.21_AD'] == '232',\n",
    "    genomatrix['22q11.21_AD'] == '000',\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['22q11.21_B_C'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "## CD\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_AD'] == '221',\n",
    "    genomatrix['22q11.21_AD'] == '223',\n",
    "    genomatrix['22q11.21_AD'] == '000',\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['22q11.21_C_D'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "\n",
    "## AC\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_AD'] == '112',\n",
    "    genomatrix['22q11.21_AD'] == '332',\n",
    "    genomatrix['22q11.21_AD'] == '000',\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['22q11.21_A_C'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "## BD\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_AD'] == '211',\n",
    "    genomatrix['22q11.21_AD'] == '233',\n",
    "    genomatrix['22q11.21_AD'] == '000',\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['22q11.21_B_D'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "## AD\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_AD'] == '111',\n",
    "    genomatrix['22q11.21_AD'] == '333',\n",
    "    genomatrix['22q11.21_AD'] == '000',\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['22q11.21_A_D'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "## note, there were some samples with 323... this is likely due to a fragmented call\n",
    "\n",
    "pd.crosstab(genomatrix['22q11.21_A_D'], genomatrix['22q11.21_C_D'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, genotype the D-H region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>22q11.21_F_H</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22q11.21_D_E</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1055</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "22q11.21_F_H     2   3\n",
       "22q11.21_D_E          \n",
       "1               14   0\n",
       "2             1055  12\n",
       "3                4   0"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions_22qDH = ['22q11.21_D_E_oneway','22q11.21_E_F_oneway', \n",
    "                 '22q11.21_F_G_oneway', '22q11.21_G_H_oneway']\n",
    "\n",
    "genomatrix['22q11.21_DH'] = (genomatrix[regions_22qDH]\n",
    "    .fillna(0)\n",
    "    .astype(int)\n",
    "    .astype(str)\n",
    "    .agg(''.join, axis=1)\n",
    "                           )\n",
    "\n",
    "## DE\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_DH'] == '1222',\n",
    "    genomatrix['22q11.21_DH'] == '3222',\n",
    "    genomatrix['22q11.21_DH'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['22q11.21_D_E'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "## EF\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_DH'] == '2122',\n",
    "    genomatrix['22q11.21_DH'] == '2322',\n",
    "    genomatrix['22q11.21_DH'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['22q11.21_E_F'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "\n",
    "## FG\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_DH'] == '2212',\n",
    "    genomatrix['22q11.21_DH'] == '2232',\n",
    "    genomatrix['22q11.21_DH'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "\n",
    "genomatrix['22q11.21_F_G'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## GH\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_DH'] == '2221',\n",
    "    genomatrix['22q11.21_DH'] == '2223',\n",
    "    genomatrix['22q11.21_DH'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "\n",
    "genomatrix['22q11.21_G_H'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "## FH\n",
    "\n",
    "conditions = [\n",
    "    genomatrix['22q11.21_DH'] == '2211',\n",
    "    genomatrix['22q11.21_DH'] == '2233',\n",
    "    genomatrix['22q11.21_DH'] == '0000'\n",
    "]\n",
    "choices = [1, 3, None]\n",
    "genomatrix['22q11.21_F_H'] = np.select(conditions, choices, default=2)\n",
    "\n",
    "pd.crosstab(genomatrix['22q11.21_D_E'], genomatrix['22q11.21_F_H'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# View table of counts for each CNV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus</th>\n",
       "      <th>n_DELs</th>\n",
       "      <th>n_DUPs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>22q11.21_A_D</td>\n",
       "      <td>501</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>16p11.2_BP4_BP5</td>\n",
       "      <td>145</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>22q11.21_F_H</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16p11.2_BP2_BP3</td>\n",
       "      <td>28</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>22q11.21_A_B</td>\n",
       "      <td>37</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2q13_NPHP1</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>22q11.21_C_D</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15q11.2_BP1_BP2</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>22q11.21_B_D</td>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>3q29</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>15q13.3_CHRNA7</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>22q11.21_D_E</td>\n",
       "      <td>14</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>22q11.21_A_C</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>16p12.2_BP2_BP3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>22q11.21_E_F</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>13q12.12</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2q21.1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>16p13.11_BP1_BP2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7q11.23_distal</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1q21.1_TAR_BP1_BP2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>10q11.21_11.23</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>7q11.23_WBS</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>17p12</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>15q13.1-13.2_BP3_BP4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>22q11.21_B_C</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1q21.1_TAR_BP2_BP3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2q13</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>22q11.21_F_G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>15q13.3_BP4_BP5_exclCHRNA7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>22q11.21_G_H</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17q12</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22q11.21_DH</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16p11.2_BP1_BP2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         locus  n_DELs  n_DUPs\n",
       "17                22q11.21_A_D     501     141\n",
       "10             16p11.2_BP4_BP5     145      73\n",
       "27                22q11.21_F_H       0      12\n",
       "1              16p11.2_BP2_BP3      28      11\n",
       "30                22q11.21_A_B      37      11\n",
       "20                  2q13_NPHP1       6      10\n",
       "9                 22q11.21_C_D      10      10\n",
       "0              15q11.2_BP1_BP2       4      10\n",
       "14                22q11.21_B_D      15       7\n",
       "25                        3q29       0       5\n",
       "11              15q13.3_CHRNA7       0       5\n",
       "28                22q11.21_D_E      14       4\n",
       "24                22q11.21_A_C      11       4\n",
       "7              16p12.2_BP2_BP3       1       3\n",
       "12                22q11.21_E_F       3       3\n",
       "26                    13q12.12       2       2\n",
       "23                      2q21.1       0       2\n",
       "32            16p13.11_BP1_BP2       4       2\n",
       "4               7q11.23_distal       0       2\n",
       "3           1q21.1_TAR_BP1_BP2       0       2\n",
       "21              10q11.21_11.23       0       1\n",
       "22                 7q11.23_WBS       0       1\n",
       "6                        17p12       0       1\n",
       "29        15q13.1-13.2_BP3_BP4       0       1\n",
       "31                22q11.21_B_C       0       1\n",
       "18          1q21.1_TAR_BP2_BP3       0       0\n",
       "19                        2q13       1       0\n",
       "15                22q11.21_F_G       0       0\n",
       "13  15q13.3_BP4_BP5_exclCHRNA7       0       0\n",
       "8                 22q11.21_G_H       0       0\n",
       "5                        17q12       1       0\n",
       "2                  22q11.21_DH       0       0\n",
       "16             16p11.2_BP1_BP2       0       0"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allloci = [a[:-7] if a[-6:] == 'oneway' else a for a in all_loci] + list(genomatrix.columns[89:])\n",
    "allloci = list(set(allloci))\n",
    "table = []\n",
    "\n",
    "for l in allloci:\n",
    "    row = {\"locus\": l}\n",
    "    if 1 in genomatrix[l].value_counts():\n",
    "        row['n_DELs'] = genomatrix[l].value_counts()[1]\n",
    "    else:\n",
    "        row['n_DELs'] = 0\n",
    "      \n",
    "    if 3 in genomatrix[l].value_counts():\n",
    "        row['n_DUPs'] = genomatrix[l].value_counts()[3]\n",
    "    else:\n",
    "        row['n_DUPs'] = 0\n",
    "    table.append(row)\n",
    "    \n",
    "counts = pd.DataFrame(table).sort_values(by='n_DUPs', ascending=False)\n",
    "pd.set_option('display.max_rows', None)\n",
    "counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "genomatrix.to_csv(f'genomatrix.recurrentCNV.{TODAY}.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
